### PR TITLE
Characters docs

### LINKED ISSUE #55598: Use the characters package in framework documentation and code samples
## Background

[The characters package](https://pub.dev/packages/characters) was recently introduced to address Dart String's limitations of manipulating extended grapheme clusters (https://github.com/dart-lang/language/issues/34). However, our user study shows that it is often difficult for developers to identify use cases where the characters package should be used instead of the Dart String API. 

## Proposal
As part of a larger plan to help developers make better API choices for text manipulation tasks, it would be helpful to suggest the characters package in the documentation of APIs that expect the developer to process text that might include extended grapheme clusters. Providing sample code that have `characters` included would be especially useful since no everyone reads the doc carefully.

Building on the [Character Library's Impact Beyonds Text Related Widgets.](https://docs.google.com/document/d/18SxWeHpds4modnvo997vIOiq01EspeFgsNvPa5ey88w/edit#) doc, I suggest the following changes:

### TextField-related
- [x] Use `characters` in the [reading values sample code](https://api.flutter.dev/flutter/material/TextField-class.html#reading-values) for the `TextField` class.
- [x] Use `characters` in [the sample code](https://api.flutter.dev/flutter/material/TextField/onChanged.html) for the `onChange` property of the `TextField` class.
- [x] Add sample code to the [API doc of TextInputFormatter](https://api.flutter.dev/flutter/services/TextInputFormatter-class.html) and use `characters` in the sample. 

### showSearch

- [x] Recommend developers use `characters` in the documentation of [showSearch](https://api.flutter.dev/flutter/material/showSearch.html) or [SearchDelegate](https://api.flutter.dev/flutter/material/SearchDelegate-class.html) and provide a sample.

### CircleAvatar
- [x] In the [Flutter Gallery demo of CircleAvatar](https://github.com/flutter/flutter/blob/master/dev/integration_tests/flutter_gallery/lib/demo/material/tabs_demo.dart), use `characters` to access the first character in a string.

###  Navigator
- [x] Provide some guidance on parsing route names that might include extended grapheme clusters. More context in [this doc](https://docs.google.com/document/d/18SxWeHpds4modnvo997vIOiq01EspeFgsNvPa5ey88w/edit#heading=h.9w7ovhhbaioc).

Feel free to create sub-issues to make this more manageable. This issue is currently blocked by https://github.com/flutter/flutter/pull/53381 and https://github.com/flutter/flutter/issues/55593.

Cc: @chunhtai

### PR DESCRIPTION
**Pull request by justinmc**
## Description

This PR aims to make our docs nudge users in the direction of handling complex characters properly, now that we include the [characters](https://pub.dev/packages/characters) package in Flutter by default (as of https://github.com/flutter/flutter/pull/53381).

## Related Issues

Closes https://github.com/flutter/flutter/issues/55598

## Tests

None, docs change only.

### CODE REVIEW

File: dev/integration_tests/flutter_gallery/lib/demo/material/tabs_demo.dart [hunk 1]
```diff
@@ -14,7 +14,7 @@ const String _kGalleryAssetsPackage = 'flutter_gallery_assets';
 class _Page {
   _Page({ this.label });
   final String label;
-  String get id => label[0];
+  String get id => label.characters.first;
```
Comment (justinmc) [inline dev/integration_tests/flutter_gallery/lib/demo/material/tabs_demo.dart:17]: I believe this example does not exist in the [gallery repo](https://github.com/flutter/gallery), so there is nothing to update over there.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 2]
```diff
@@ -131,6 +131,8 @@ class _CupertinoTextFieldSelectionGestureDetectorBuilder extends TextSelectionGe
 /// field (e.g., by pressing a button on the soft keyboard), the text field
 /// calls the [onSubmitted] callback.
 ///
+/// {@macro flutter.widgets.editableText.complexCharacters}
```
Comment (InMatrix) [inline packages/flutter/lib/src/cupertino/text_field.dart:134]: What does this macro do?

File: packages/flutter/lib/src/material/text_field.dart [hunk 3]
```diff
@@ -258,6 +258,14 @@ class _TextFieldSelectionGestureDetectorBuilder extends TextSelectionGestureDete
...
 /// Keep in mind you can also always read the current string from a TextField's
 /// [TextEditingController] using [TextEditingController.text].
+///
+  /// ## Handling emojis and other complex characters
+/// {@macro flutter.widgets.editableText.complexCharacters}
+///
+/// In the live Dartpad example above, try typing the emoji 👨‍👩‍👦
+/// into the field and submitting. Because the example code measures the length
+/// with `value.characters.length`, the emoji is correctly counted as a single
+/// character.
```
Comment (InMatrix) [inline packages/flutter/lib/src/material/text_field.dart:268]: Nice!

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_field.dart:134]: Elsewhere I defined a template with:

```
{@template flutter.widgets.editableText.complexCharacters}
```

So now every place that uses this macro will import the text defined in the template.

**Review by InMatrix — APPROVED** [review]
Thanks for including the Characters API in the these docs. Those changes look good to me.

**Comment by InMatrix** [comment]
By the way, we should probably update some cookbook entries linked from these API docs as well, for example, https://flutter.dev/docs/cookbook/forms/text-field-changes.

**Comment by justinmc** [comment]
@InMatrix No problem, thanks for the clear issue and design doc!

Do you know how to update those cookbook pages?  Are they in another repo somewhere?

**Comment by InMatrix** [comment]
@justinmc The cookbook is in the website repo. I think the initial version was created by @brianegan. @RedBrogdon would know how to update text-related entries. Let me know if you'd like me to file a bug for that.

**Comment by justinmc** [comment]
@InMatrix Yes please, can you file the issue and tag me?

**Comment by InMatrix** [comment]
Here you go: https://github.com/flutter/website/issues/4834

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 4]
```diff
@@ -946,6 +946,22 @@ class EditableText extends StatefulWidget {
   /// {@end-tool}
   /// {@endtemplate}
   ///
+  /// ## Handling emojis and other complex characters
+  /// {@template flutter.widgets.editableText.complexCharacters}
+  /// It's important to always use
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:951]: I think that's probably not true If the editable text is only meant for english alphabets and/or numbers (verification code, for example), and there's a formatter that guarantees that?

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/text_field.dart:134]: Should this paragraph have the same   /// ## Handling emojis and other complex characters header as well?

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 4 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/editable_text.dart:951]: True.  I'll change it to something like "when working with user input text that may contain complex characters"

**Pull request merged**